x/oracle/keeper/keeper.go

```diff
@@ -361,3 +361,72 @@ func (k Keeper) ValidateFeeder(ctx sdk.Context, feederAddr sdk.AccAddress, valid
 
 	return nil
 }
+
+func (k Keeper) GetPriceSnapshots(ctx sdk.Context) types.PriceSnapshotHistory {
+	store := ctx.KVStore(k.storeKey)
+	snapshotsBytes := store.Get(types.GetPriceSnapshotsKey())
+	if snapshotsBytes == nil {
+		return types.PriceSnapshotHistory{}
+	}
+
+	priceSnapshots := types.PriceSnapshotHistory{}
+	k.cdc.MustUnmarshal(snapshotsBytes, &priceSnapshots)
+	return priceSnapshots
+}
+
+func (k Keeper) SetPriceSnapshots(ctx sdk.Context, snapshots types.PriceSnapshots) {
+	// shouldn't be used directly, use "add" instead for individual price snapshots
+	store := ctx.KVStore(k.storeKey)
+	priceSnapshots := types.PriceSnapshotHistory{PriceSnapshots: snapshots}
+	bz := k.cdc.MustMarshal(&priceSnapshots)
+	store.Set(types.GetPriceSnapshotsKey(), bz)
+}
+
+func (k Keeper) AddPriceSnapshot(ctx sdk.Context, snapshot types.PriceSnapshot) {
+	params := k.GetParams(ctx)
+	lookbackDuration := params.LookbackDuration
+	snapshotHistory := k.GetPriceSnapshots(ctx)
+	snapshots := snapshotHistory.PriceSnapshots
+	snapshots = append(snapshots, snapshot)
+	var indexToSliceBefore int
+	indexToSliceBefore = 0
+	for i, snapshot := range snapshots {
+		if snapshot.SnapshotTimestamp+lookbackDuration >= ctx.BlockTime().Unix() {
+			// we want to keep one data point that is outside of the lookback duration
+			// so that we can calculate the TWAP for the first interval
+			indexToSliceBefore = i - 1
+			if indexToSliceBefore < 0 {
+				// if there is no datapoint outside of range, we simply keep all of data
+				indexToSliceBefore = 0
+			}
+			break
+		}
+	}
+	snapshots = snapshots[indexToSliceBefore:]
+	k.SetPriceSnapshots(ctx, snapshots)
```

Comment:
we would be writing the entire price history to disk each time we add a new snapshot, which might be a bit too heavy. Maybe we can use timestamp as a subkey so that each snapshot can be added/deleted individually?

Reply:
ah yeah, this is a good point, I'll make this change